### ISSUE TITLE
Inefficient `moveaxis` in `dot_general` batching rule

### ISSUE THREAD

**Issue by hexahedria** [google/jax#2972]
When batching a `dot_general` primitive, the batch axis is [moved to the end](https://github.com/google/jax/blob/b543652332a98e5c15db04651417404513aa4d1a/jax/lax/lax.py#L2563) of the array. Right now, this can cause problems for high-rank tensors because the order of axes affects the way XLA compiles the operation (in particular, can result in padding out the moved batched axis to a multiple of 128, instead of using one of the other larger axes).

Instead of moving the batch axis to the end, it should be possible to keep it in its original location, which would give the user more control over how XLA compiles the computation.

A repro that produces this behavior:
```
x = jnp.zeros((16, 3, 256, 256))
y = jnp.zeros((2, 4, 256))
z = jnp.zeros((16, 3, 4, 256, 2048))

def go(x, y, z):
  return jnp.einsum("vnn,sn,vsnj->nj", x, y, z)

@jax.jit
def repro(x, y, z):
  go_bat = jax.vmap(jax.vmap(go, (0, None, 0)), (None, 0, None))
  return go_bat(x, y, z)

repro(x, y, z)
```
